### PR TITLE
Imviz: Load 3D slices, enable delayed_linking

### LINKED ISSUE #1055: Imviz should provide API to batch load a list of images with delayed linking
**Reporter:** pllim

As @orifox found out, if you run `imviz.load_data(...)` in a loop, it gets progressively slower because it tries to re-link the entire data collection at every iteration. Imviz should provide an API to batch load a list of images with delayed linking.

* Idea 1: `imviz.batch_load_data(list_of_images)` to load everything at once.
* Idea 2: `imviz.load_data(individual_image, no_linking=True)` to load one at a time yourself but you have to remember to run `imviz.load_data(last_image, no_linking=False)` in the last iteration.
* Idea 3: `imviz.load_cube(data_cube, axis=n)` to display each spatial slice as an image, which was what Ori was actually trying to do. p.s. If Cubeviz can load arbitrary cubes, this idea might be moot anyway. Also see #1040

[🐱](https://jira.stsci.edu/browse/JDAT-2110)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### LINKED ISSUE #1058: Imviz bypassed ndim check for AsdfInFits data
**Reporter: Pey-Lian Lim**

[~cpacifici] was able to load a 4D cube into Imviz (on the released version, so this is without https://github.com/spacetelescope/jdaviz/pull/1056 ) because the UNCAL file has ASDF-in-FITS, so it bypassed a dimension check. While it loaded, the viewer does not tell you what you are actually seeing on the 2D display and coordinates info only show X/Y and no value, so the viewer is not very useful anyway.

Proposed solution:

1. Add ndim check in ASDF parser to throw error for ndim!=2.
2. Ask user to use new feature added in https://github.com/spacetelescope/jdaviz/pull/1056 to load this file instead.

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to enable:

* Loading a data cube as individual 2D slices at `axis=0`.
* Enable option to disable linking at every iteration when user wants to call `imviz.load_data(...)` in a loop.

Fixes #1055

Fixes #1058

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by kecnry** [comment]
Is there any reason a user would intentionally leave data unlinked?  Could we maybe add checks at other interactive steps (creating a new viewer, blinking, creating/editing subregions, etc) that would handle the linking step if they ever did use `do_link=False` but then forget to manually link or will that apply to too many different places and be hard to maintain?

**Comment by pllim** [comment]
> Is there any reason a user would intentionally leave data unlinked?

Intentionally, no. Because blinking does not work if data are not linked.

I think it would add unnecessary complexity to try to be too smart and auto perform the linking. My take is by the time someone wants to load stuff like this in a loop into Imviz, hopefully they are power-user that know manual linking is needed. If they don't, then it is not the end of the world, they manually run one more line of code or use the "Imviz Links Control" plugin (though with the latter, I think you have to uncheck and recheck all the data in Data dropdown, not sure why that is).

**Review by kecnry — APPROVED** [review]
I don't have an example case to test against (outside the notebook), but code changes all seem reasonable given the conversation above.

File: jdaviz/configs/imviz/tests/test_parser.py [hunk 1]
```diff
@@ -88,6 +88,23 @@ def test_parse_numpy_array(self, imviz_app):
...
 
+    def test_parse_numpy_array_3d(self, imviz_app):
+        # This data has values in axis=0 that correspond to axis num.
+        n_slices = 5
+        slice_shape = (2, 3)
+        arr = np.stack([np.zeros(slice_shape) + i for i in range(n_slices)])
+
+        # We use higher level load_data() here to make sure linking does not crash.
+        imviz_app.load_data(arr, data_label='my_slices')
+        assert len(imviz_app.app.data_collection) == n_slices
```
Comment (kecnry) [inline jdaviz/configs/imviz/tests/test_parser.py:99]: is it worth adding testing of the for-loop deferred linking use case?

Comment (pllim) [reply jdaviz/configs/imviz/tests/test_parser.py:99]: Yes, good call. I'll add that!

**Comment by pllim** [comment]
> I don't have an example case

Working on a fun example. Stay tuned!

File: jdaviz/configs/imviz/tests/test_parser.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/imviz/tests/test_parser.py]: Test added.

**Comment by pllim** [comment]
No idea what the `RuntimeError: Unable to locate core schemas` is about. I asked @eslavich .

**Comment by pllim** [comment]
@camipacifici indicated that we can merge this first while POs think about what does it mean to load `>2D` data *file* into a 2D viewer. Just need one more approval. 🙇‍♀️

**Review by ibusko — APPROVED** [review]
Looks good to me. The demo works as expected. I just wonder what would happen if someone loads a cube with many slices, like a spectrogram with 1000 wavelength values or something like that. But I think it should be merged now.

**Comment by pllim** [comment]
I'll open a follow-up PR to add warning to the docstring about performance.

**Comment by pllim** [comment]
Thanks for the reviews!

**Pull request merged**